architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+This "lazy loading" optimizes pod startup time, for the common case where an application is not sending outbound traffic instantly.
+For applications that do, they will see a slight latency increase during startup; most applications can handle this far better than failed connections.
+
+## Pod Shutdown
+
+### Pod Shutdown Requirements
+
+During Pod shutdown, we MUST meet the constraint that the Pod shutting down can send traffic up until its full stopped.
+Pod shutdown can be a long process; while `terminationGracePeriodSeconds` is typically ~30s, it can be extremely high, leaving terminating pods running for extended periods of time with an expectation of networking.
+
+Additionally, we MUST accept incoming traffic during this time, both new and existing.
+Typically, it is rare to get new connections after shutdown has started - once shutdown is triggered the endpoint is marked `NotReady` and removed from load balancing consideration.
+However, some applications do not utilize Service readiness (connecting directly to the pod, etc).
+Additionally, eventually consistency means we will often get some final requests trickle in before all clients are aware the Pod is shutting down.
+
+Finally, we SHOULD tell our (HBONE) peers we are shutting down.
+This can help ensure that clients do not hold onto pooled connections for too long.
+This is not strictly required; Ztunnel clients have both keepalives (which will start to fail once the other end is shut down) and pool idle timeouts (which will close the connection when it has had no activity for some period; because the backend is gone, there will be no activity).
+However, it would be ideal to give immediate feedback rather than relying on these mechanisms.
+
+### Pod Shutdown Implementation
+
+[ZDS](../../pkg/zdsapi/zds.proto) exposes a `DelWorkload` message to signal to Ztunnel that a pod has been terminated.
+This can be triggered either by a Pod being actually deleted, or has transitioned to a [terminal phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/).
+At this point, the pod and all containers have fully shut down, and we can tear everything down.
+
+Because the application(s) are already entirely gone, there is no need to have any "draining" time -- we can immediately terminate.
+Furthermore, because the Pod is torn down, we cannot even send any data at all (at least on most clusters, the `veth` is torn down upon Pod deletion).
+Notably, this prevents us from sending a GOAWAY to notify peers we have shutdown.
+
+As a result, while we meet our two must-haves (send and accept traffic throughout the entire pod lifetime), we do not have an optimal solution here yet.
+Perhaps hooking into the [CNI `DEL`](https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications) could be a path forward here.
```

Comment:
```suggestion
```

This probably just belongs in an issue where the problem can be discussed, not this doc - the hook (or lack thereof) isn't really the problem here AFAIK, it's that you can't GOAWAY clients that aren't aware they're being proxied.

Reply:
No, that is not really the problem I am trying to convey. The problem is HBONE clients, which we know can GOAWAY, cannot because we cannot send it at the right time. This isn't talking about the fact we are handling opaque TCP streams.

I will remove this CNI DEL part though, agree with your comment